spec/02-integration/09-hybrid_mode/08-lazy_export.lua

```diff
@@ -0,0 +1,139 @@
+local helpers = require "spec.helpers"
+
+local admin_client
+
+local function cp(strategy)
+  assert(helpers.start_kong({
+    role = "control_plane",
+    cluster_cert = "spec/fixtures/ocsp_certs/kong_clustering.crt",
+    cluster_cert_key = "spec/fixtures/ocsp_certs/kong_clustering.key",
+    database = strategy,
+    cluster_listen = "127.0.0.1:9005",
+    nginx_conf = "spec/fixtures/custom_nginx.template",
+    -- additional attributes for PKI:
+    cluster_mtls = "pki",
+    cluster_ca_cert = "spec/fixtures/ocsp_certs/ca.crt",
+  }))
+  admin_client = assert(helpers.admin_client())
+end
+
+local n = 0
+local function touch_config()
+  n = n + 1
+  assert(admin_client:send({
+    method = "POST",
+    path = "/services",
+    body = {
+      name = "test" .. n,
+      host = "localhost",
+    },
+    headers = {
+      ["Content-Type"] = "application/json",
+    },
+  }))
+end
+
+local function old_dp()
+  assert(helpers.start_kong({
+    role = "data_plane",
+    legacy_hybrid_protocol = true,
+    database = "off",
+    prefix = "dp1",
+    cluster_cert = "spec/fixtures/ocsp_certs/kong_data_plane.crt",
+    cluster_cert_key = "spec/fixtures/ocsp_certs/kong_data_plane.key",
+    cluster_control_plane = "127.0.0.1:9005",
+    proxy_listen = "0.0.0.0:9002",
+    -- additional attributes for PKI:
+    cluster_mtls = "pki",
+    cluster_server_name = "kong_clustering",
+    cluster_ca_cert = "spec/fixtures/ocsp_certs/ca.crt",
+  }))
+end
+
+local function wrpc_dp()
+  assert(helpers.start_kong({
+    role = "data_plane",
+    database = "off",
+    prefix = "dp2",
+    cluster_cert = "spec/fixtures/ocsp_certs/kong_data_plane.crt",
+    cluster_cert_key = "spec/fixtures/ocsp_certs/kong_data_plane.key",
+    cluster_control_plane = "127.0.0.1:9005",
+    proxy_listen = "0.0.0.0:9003",
+    -- additional attributes for PKI:
+    cluster_mtls = "pki",
+    cluster_server_name = "kong_clustering",
+    cluster_ca_cert = "spec/fixtures/ocsp_certs/ca.crt",
+  }))
+end
+
+
+for _, strategy in helpers.each_strategy() do
+  describe("lazy_export with #".. strategy, function()
+    describe("no DP", function ()
+      setup(function()
+        cp(strategy)
+      end)
+      teardown(function ()
+        helpers.stop_kong()
+      end)
+      it("test", function ()
+        touch_config()
+        assert.logfile().has.no.line("exporting config for legacy protocol", true)
```

Comment:
The negative error.log checks always make me nervous, because it's so easy to for the test to pass improperly if the log entry we're looking for simply hasn't been logged _yet_, or if a future, unrelated commit changes the log message ever so slightly such that our assertion would never match it.

So if there's anything from #9369 that I think is worth making the case to bring over to this PR it would be the logging+testing changes. Adding/updating some ngx.log calls allowed me to turn all of the negative assertions to positive ones. WDYT?

Reply:
@flrgh Could you make the logging change? As long as we are not spamming the log files with the change I'm all open for it.

Reply:
done.